### ISSUE TITLE
mshadow_op square_root for half_t

### ISSUE THREAD

**Issue by Godricly** [dmlc/mxnet#2478]
The function within square_root is the sqrtf. Is this one compatible with half_t type?
I got compiling problem when trying to get batch-norm for half type.
on [line 103](https://github.com/dmlc/mxnet/blob/master/src/operator/batch_norm-inl.h#L103). It has some problem to derive a half_t type output, even if I casted the eps into DType.

**Comment by sxjscience** [comment]
How about changing it to `sqrt`?

**Comment by vchuravy** [comment]
We might have to add fallback for mathematical functions for `half_t`

I can work on that.

**Comment by vchuravy** [comment]
@Godricly can you put up the code?

**Comment by Godricly** [comment]
I'll try. I'm using MxNet in an intranet. So it takes some time. I'll let you when I'm ready.

**Comment by vchuravy** [comment]
You can also make a patch and send it to me via e-mail

**Comment by Godricly** [comment]
@vchuravy Is there anyway to bypass this problem on cpu side?

**Comment by vchuravy** [comment]
I have some ideas that I will work on tomorrow. I have to reproduce the
specific problem first.

**Comment by Godricly** [comment]
@vchuravy i Just created a [branch](https://github.com/Godricly/mxnet/tree/BN-DType/src/operator) so u can start from here. They are some parameters were float type in previous BN. Should we convert them into DType? I left commets started with "Issue: " so u can jump there by find.

**Comment by Godricly** [comment]
actually, The fc layer does not support FP16 on cpu side.
[code link](https://github.com/dmlc/mxnet/blob/master/src/operator/fully_connected.cc#L19)

**Comment by vchuravy** [comment]
@Godricly you nearly had everything, the primary thing that was missing is that all constant need to be `tcast<DType>(scalar(...))` in expressions.

**Comment by Godricly** [comment]
I was thinking about it. I had them converted in my  intranet code. And I created the test code following test_operator script and it passed test. I'll updated them tomorrow and check with cudnnv4. One issue here is that test_operator will not test self implemented batch norm.

**Comment by vchuravy** [comment]
Was your problem with `square_root` a compile time or runtime issue?

**Comment by Godricly** [comment]
it was compiling issue with those mshadow_op on cpu side using half_t. But now In batchnorm.cc, the half_t is disable for cpu like it is in fc layer. So it passed.
